File: epi_judge_java/epi/IntSquareRoot.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class IntSquareRoot {

    @EpiTest(testDataFile = "int_square_root.tsv")
    public static int squareRoot(int k) {
        if (k < 0) return -1;
        long left = 0;
        long right = (long) k + 1;
        while (left + 1 < right) {
            long mid = left + (right - left) / 2;
            if (mid * mid > k) right = mid;
            else left = mid;
        }
        return (int) left;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IntSquareRoot.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/int_square_root.cc
#include "test_framework/generic_test.h"

int SquareRoot(int k) {
    if (k < 0) return -1;
    int left = -1;
    int right = 1e9;
    while (left + 1 < right) {
        int candidate = left + (right - left) / 2;
        long long sq = (long long) candidate * candidate;
        if (sq > k) {
            right = candidate;
        } else {
            left = candidate;
        }
    }
    return left;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"k"};
    return GenericTestMain(args, "int_square_root.cc", "int_square_root.tsv",
                           &SquareRoot, DefaultComparator{}, param_names);
}
